File: src/main/java/com/kishlaly/ta/analyze/tasks/blocks/one/Long_ScreenOne_EMA_ThreeAscending.java
package com.kishlaly.ta.analyze.tasks.blocks.one;

import com.kishlaly.ta.model.BlockResult;
import com.kishlaly.ta.model.SymbolData;
import com.kishlaly.ta.model.indicators.EMAJava;
import com.kishlaly.ta.model.indicators.Indicator;
import com.kishlaly.ta.utils.CollectionsTools;
import com.kishlaly.ta.utils.Log;

import java.util.List;

import static com.kishlaly.ta.analyze.BlockResultCode.OK;
import static com.kishlaly.ta.analyze.BlockResultCode.THREE_EMA_NOT_ASCENDING_SCREEN_1;

/**
 * the last three EMA are increasing
 */
public class Long_ScreenOne_EMA_ThreeAscending implements ScreenOneBlock {
    @Override
    public BlockResult check(SymbolData screen) {
        List<EMAJava> screen_1_EMA26 = (List<EMAJava>) screen.indicators.get(Indicator.EMA26);
        EMAJava ema3 = CollectionsTools.getFromEnd(screen_1_EMA26, 3);
        EMAJava ema2 = CollectionsTools.getFromEnd(screen_1_EMA26, 2);
        EMAJava ema1 = CollectionsTools.getFromEnd(screen_1_EMA26, 1);

        boolean ascending = ema3.getValue() < ema2.getValue() && ema2.getValue() < ema1.getValue();

        if (!ascending) {
            Log.recordCode(THREE_EMA_NOT_ASCENDING_SCREEN_1, screen);
            Log.addDebugLine("Three values of the EMA do not grow on the long-term screen");
            return new BlockResult(screen.getLastQuote(), THREE_EMA_NOT_ASCENDING_SCREEN_1);
        }
        return new BlockResult(screen.getLastQuote(), OK);
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/tasks/blocks/one/Long_ScreenOne_EMA_ThreeAscending.kt
package com.kishlaly.ta.analyze.tasks.blocks.one

import com.kishlaly.ta.model.BlockResult
import com.kishlaly.ta.model.BlockResultCode
import com.kishlaly.ta.model.SymbolData
import com.kishlaly.ta.model.indicators.EMA
import com.kishlaly.ta.model.indicators.Indicator
import com.kishlaly.ta.utils.CollectionUtils
import com.kishlaly.ta.utils.Log

/**
 * the last three EMA are increasing
 */
class Long_ScreenOne_EMA_ThreeAscending : ScreenOneBlock {
    override fun check(screen: SymbolData): BlockResult {
        val screen_1_EMA26 = screen.indicators[Indicator.EMA26] as List<EMA>
        val ema3 = CollectionUtils.getFromEnd<EMA>(screen_1_EMA26, 3)
        val ema2 = CollectionUtils.getFromEnd<EMA>(screen_1_EMA26, 2)
        val ema1 = CollectionUtils.getFromEnd<EMA>(screen_1_EMA26, 1)

        val ascending = ema3.value < ema2.value && ema2.value < ema1.value

        if (!ascending) {
            Log.recordCode(BlockResultCode.THREE_EMA_NOT_ASCENDING_SCREEN_1, screen)
            Log.addDebugLine("Three values of the EMA do not grow on the long-term screen")
            return BlockResult(screen.lastQuote, BlockResultCode.THREE_EMA_NOT_ASCENDING_SCREEN_1)
        }

        return BlockResult(screen.lastQuote, BlockResultCode.OK)
    }
}